### ISSUE TITLE
Proposed standardised ValidityForMutables

### ISSUE THREAD

**Issue by montanajava** [blw-ofag-ufag/eCH-0261#26]
Create an reusable element called ValidityForMutables.

This is very similar to:

https://github.com/blw-ofag-ufag/eCH-0261/issues/19

with the following changes:

- iteration of type integer added.
- supersededById becomes supersededByIteration
- supersedesId becomes supersedesIteration.

The name of the element is of course not great: Ideas welcome.

This does not _replace_ Validity, rather it is proposed as an addition thereto.

**Architectural Considerations**
The motivation of this class is that using Validity for mutable classes includes a misnomber: An instance of a mutable class does not supersede an "Id" but in fact an iteration of the same id. 

The word iteration was chosen because in many programming languages or frameworks the word "version" is already reserved, or taken or occupied. "version" in those settings refers to a technical version, whereas here, "iteration" is most definitely a domain-specific attribute.

**Comment by montanajava** [comment]
If this proposal is accepted, then references to Validity where an additional "version" was heretofore used, would have to be updated, e.g., eCH-0263, ManureAndRecyclingProduct (sic).

**Comment by AFoletti** [comment]
I am not sure I understand how this can work and how it is different from what already done and discussed in #19 

If you have the very same object (same ID) it can only exist in one iteration at any given time. If you have multiple iterations of the same object coexisting in time _they have to be actually different objects_. Proof of that is that you can only differenciate them by ID+iteration value, actually building a new composite ID.

The "iteration" attribute has in my opinion another issue: what if an object needs diverging histories?

1. Object 1 (iteration 0) is defined by organization A
2. Object 1 is updated by organization A (iteration 1)
3. Organization B creates yet a new version of Object 1, based on iteration 0. Is it iteration 1 or 2? Formally 1, since based on 0 and not the "old" 1.

To sum up:

- I am under the impression that any kind of "iteration" attribute solves very little and opens the door to data-chaos. Historisation is already solved by #19 and the concept of "iteration" is actually a derived information rather than a primary attribute
- If we really want to implement an iteration attribute, I would not do it in a new class but make it optional in the class already defined in #19 

My proposal is to leave it be and keep the class as discussed in #19, without any iteration info

**Comment by montanajava** [comment]
The motivator was PSM data, which does in fact mutate over time. I agree that an iteration without a change in ID is, in effect, a composite key.

Well architected systems do not do what you described in diverging histories. A consuming system of PSM data would thus not change data but rather forward any required change to the source system for the change to be made there.

I just felt that the present Validity class did not cover cleanly this PSM edge case.  But probably such an edge case does not belong in a standard at all! Thanks for taking a look.  From my point of view, you can close it if with a resolution “won’t fix.”

**Comment by AFoletti** [comment]
> Well architected systems do not do what you described in diverging histories. A consuming system of PSM data would thus not change data but rather forward any required change to the source system for the change to be made there.
> 

I see your point and I 100% agree with you.

Nevertheless, you put forward a _system_ (as in "technical implementation") example, while my reasoning was more at the conceptual level.

I would be interested in other opinions and will let @lars-steffen close the ticket with the chosen resolution.

**Comment by AFoletti** [comment]
Closing this as part of the "end of 2024" pruning effort. Please reopen if really needed and real work is planned.
